unic/utils/src/char_property.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2017 The UNIC Project Developers.
+//
+// See the COPYRIGHT file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+//! Taxonomy and Contracts for Character Property types.
+//!
+//! See also the list of types of character properties defined in the UCD:
+//! <http://unicode.org/reports/tr44/#About_Property_Table>, in [Unicode® Standard Annex #44 —
+//! Unicode Character Database](http://unicode.org/reports/tr44/#About_Property_Table)
+
+
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+
+/// A Character Property defined on all characters.
+///
+/// Examples: *Age*, *Name*, *General_Category*, *Bidi_Class*
+pub trait CharProperty
```

Comment:
This should probably extend from `OptionCharProperty`, as do `Eq : PartialEq` and `Ord : PartialOrd`.

I would then do a blanket impl from `CharProperty` to `OptionCharProperty`:

```rust
impl<P: CharProperty> OpionCharProperty for P {
    fn of(ch: char) -> Option<Self> {
        Some(<Self as CharProperty>::of(ch))
    }
}
```

[Really rough playground proof of concept](https://play.rust-lang.org/?gist=25da653487d1713a01d1aa0f83773c6d&version=stable)